### ISSUE TITLE
Recommend addins to start targeting `net5.0` (in addition to `netstandard2.0` and `net461`)

### ISSUE THREAD

**Issue by augustoproiete** [cake-build/website#1513]
With the Cake v1.0 release targeting `net5.0` we should also recommend addins to multi-target to `net5.0` as well, given that it supersedes `netstandard2.0`.

I believe starting with v1.0 we should recommend addins to target `net461`, `netstandard2.0` and `net5.0` for broader reach.

Then in the future (after `net6.0`) we can re-evaluate.

----

- [Best practices for writing addins](https://cakebuild.net/docs/extending/addins/best-practices#net-target-version)

**Comment by pascalberger** [comment]
@augustoproiete What advantages do you expect from additionally targeting `net5.0` if addin doesn't use any specific code outside of `netstandard2.0`?

**Comment by augustoproiete** [comment]
The main reason is to encourage addin authors to keep their code up-to-date with the latest version of .NET, and at the same time to follow the [guidelines from Microsoft](https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/#what-you-should-target):

> - Use `netstandard2.0` to share code between .NET Framework and all other platforms.
> - Use `netstandard2.1` to share code between Mono, Xamarin, and .NET Core 3.x.
> - Use `net5.0` for code sharing moving forward.

Indirectly, this helps us in the long run to (more easily) drop the `netstandard2.0` target in `Cake.Core` and `Cake.Common` in a future major version, so that these libraries can leverage new features of `net5.0` with a lower cost of maintenance for us (i.e. using `#if` is ugly and knowing when you can and can't use certain features (+ polyfill) adds a lot of complexity).

**Comment by pascalberger** [comment]
> The main reason is to encourage addin authors to keep their code up-to-date with the latest version of .NET, and at the same time to follow the [guidelines from Microsoft](https://devblogs.microsoft.com/dotnet/the-future-of-net-standard/#what-you-should-target):

Our use case is not really handled in these guidelines: Addins are different than libraries, since they are directly bound to Cake. Cake itself is different to app components, since it is a tool running on different platforms.

Dropping `netstandard2.0` and going from  `net461`, `netstandard2.0` and `net5.0` to `net461` and `net5.0` won't really help addin authors, since you stuck more or less in the same situation. Getting rid of Cake runner for .NET Framework and .NET Core and to only target `net5.0` would be a different story, but I don't see that to suggest to multi-target to `net5.0` before is of any help. On the contrary, as long as an addin won't make use of `net5.0` specific APIs it will just increase addin size.

**Comment by augustoproiete** [comment]
Don't skip my first point, though which is the most important! :smile::smile:

> **The main reason is to encourage addin authors to keep their code up-to-date with the latest version of .NET**

You know better than I do how hard it is for addin maintainers to keep them up-to-date and the Cake team has been jumping in and taking on the burden of getting them up-to-speed (@nils-a I'm looking at you :sweat_smile:)... If we set the bar higher on the guidelines, there's a better chance our own jobs will be easier if/when we have to help support addins if the author already did the work necessary to target more modern versions of .NET.

I don't disagree with the fact that it will increase the size of the NuGet package - that's true, but I think it's a small price to pay, where the benefits outweigh the cost in the long run.

So as a principle, I think the recommendation should be to always target the latest and greatest .NET version available, and multi-target to older versions for broader reach.

**Comment by pascalberger** [comment]
> Don't skip my first point, though which is the most important! 😄😄
> 
> > **The main reason is to encourage addin authors to keep their code up-to-date with the latest version of .NET**
> 

But for what advantage? You can't make use of any of the additional features as long as you still multi-target to the platforms Cake supports or you need to use conditional code. 

Promoting multi-targeting to `net5.0` for 99% of addin maintainers just adds additional work, increases file size, and adds the risk that they start to only target `net5.0` which will end in a much worse user experience. IMHO guidelines should promote what's best and easiest for most of the addins and provides best compatibility for users, which is still multi-targeting `net461` and `netstandard2.0` as long as we don't change TFM in Cake. Even if we once drop everything beside `net5.0`, addins targeting `netstandard2.0` will still continue to work. If an addin author wants to take advantage of newer features, he still can multi-target, guidelines are just guidelines and not technical requirements, but we shouldn't promote this for every addin author. 
If you look at what kind of addins exist, then there are the simple addins which just wrap a tool (which are the majority), for which there's probably not much need for using of new features and multi-targeting will end with same code for every TFM and just increase file size.
Then there are more complex addins which contain more code (like Cake.Git or Cake.Issues.Reporting.Generic, etc) , where every additional TFM is just pure pain to maintain, lot of work to test and support (it's not only the TFM you need to test and support, but the combination of TFM and different runners). Promoting an additional TFM in these cases, just for the sake of targeting an additional TFM, without bringing real value is not something we should do.

**Comment by augustoproiete** [comment]
I look at the guidelines more with a view of "new" addins that will be developed, and I think you're looking at the guidelines more with a view of "existing" addins. Is that fair to say?

Do you think it would be worth splitting the guidelines in two parts and we can probably agree on different requirements for new vs existing addins?

- Guidelines for developing new addins after v1.0 ships
- Guidelines for maintaining existing addins after v1.0 ships

### Guidelines for developing new addins after v1.0 ships

My view is that `netstandard2.0` is dead, therefore recommending that **new** code should target `netstandard2.0` and `net461` **_only_** doesn't make sense to me. New code should (at a minimum) target the latest version available at the time that you do File -> New Project (that's the principle I'm trying to convey).

I think either we recommend **new** addins to be `net5.0` at a minimum, or we recommend that **new** addins multi-target `net5.0` and `netstandard2.0` and `net461` - which is what I am proposing.

Putting my addin author hat on :billed_cap:, I would _start_ with targeting `net5.0` for my addin, and if the cost of multi-targeting to other versions such as `netstandard2.0` and `net461` is low, then great I'll do it, otherwise no, and it'll be `net5.0` only...

The other way around is worse (IMO): Start with `netstandard2.0` and `net461` and later discover that moving to `net5.0` is harder and not do it and abandon the addin.


### Guidelines for maintaining existing addins after v1.0 ships

I'm less concerned about this one because (_and I agree with you that many, if not all, of the existing addins that today target `netstandard2.0` will continue to work just fine without having to target `net5.0`_):

1. These are just recommendations and as we know very few addin authors follow the recommendations... We have the data to prove it: As of December 20th 2020, only `25` out of `272` (10%) addins follow the recommendation of targeting both `netstandard2.0` and `net461` ([source](https://github.com/cake-build/website/pull/1400#issue-543169745)). So I expect the impact of recommending `net5.0` to be low.

2. I do see value as I said in my previous comment, in keeping the projects active and healthy. Putting my build engineer hat on now :billed_cap: I'm more likely to take a dependency on an addin that had the last commit 3 months ago, than if the last commit was 3 years ago. If I see someone updating their addins to keep up with the latest versions of Cake and/or .NET, it sends me a signal that the addin is still active and the maintainer is engaged.

---

> Promoting multi-targeting to net5.0 [...] and adds the risk that they start to only target `net5.0`

You say that like it's a bad thing. IMHO we should be encouraging devs to move forward and leave the .NET Framework runner behind, leave the CoreCLR runner behind, and make the jump to the future: The dotnet tool running on .NET 5.

At the end of the day, we're talking about build tools... Not applications, and as you know, one can easily have a build pipeline running on .NET 5, but building applications that still target .NET Framework 4.0 or even older... i.e. The build pipeline doesn't have to be as old as the application being built... It can evolve much faster than apps, which leaves little reason not to upgrade it.

Even Microsoft dev teams (with paid employees) dropped `netstandard2.0` moving forward (e.g. EF Core), so I don't think it's reasonable to expect unpaid volunteers (like addin authors) to keep supporting it indefinitely.

**Comment by pascalberger** [comment]
TL;DR: Addin guidelines should provide minimum requirements, match platform support of Cake and ensure best user experience. Every author is free to add more if preferred or necessary.

> I look at the guidelines more with a view of "new" addins that will be developed, and I think you're looking at the guidelines more with a view of "existing" addins. Is that fair to say?
> 
> Do you think it would be worth splitting the guidelines in two parts and we can probably agree on different requirements for new vs existing addins?

For me it doesn't make a difference if a new addin is written or an existing addin is maintained and I don't see any reason why we should have different guidelines and best practices depending on when development of an addin has started. We also did create some automated PRs in the past to addins for best practices. It would be very weird to have different best practices for TFM applied to addins depending if they are made before or after a certain day, where a Cake version was released, where the specific Cake version itself didn't change the TFM.

> My view is that `netstandard2.0` is dead, therefore recommending that **new** code should target `netstandard2.0` and `net461` **_only_** doesn't make sense to me. New code should (at a minimum) target the latest version available at the time that you do File -> New Project (that's the principle I'm trying to convey).
> 
> I think either we recommend **new** addins to be `net5.0` at a minimum, or we recommend that **new** addins multi-target `net5.0` and `netstandard2.0` and `net461` - which is what I am proposing.

That's true. But the thing is that Cake addins are not new standalone projects, but very directly bound to Cake. Changes to supported Framework versions should be lead by Cake and we should not push the addin ecosystem in a direction where we suggest to support only a subset of what Cake itself supports. This would be a nightmare for users. In the past it was always Cake which lead the direction (from .NET Framework to different versions of .NET Core) and we're now at a point where most of the addins work with most of the runners and TFM. Having official suggestion that addins should no longer support certain runners at all would bring absolutely no benefit and just disrupt the ecosystem.

> Putting my addin author hat on 🧢, I would _start_ with targeting `net5.0` for my addin, and if the cost of multi-targeting to other versions such as `netstandard2.0` and `net461` is low, then great I'll do it, otherwise no, and it'll be `net5.0` only...

It's up for any addin author to decide for himself if it brings any value to target `net5.0` or not. But we definitely should not officially recommend addin authors to not support all Cake runners. We had this discussion for Cake itself and decision was to release 1.0 with the current TFM. For Cake 2.0 we can discuss TFM. But this should be done for Cake first and from there find its way into addin guidelines.

We have the idea to show how well an addin follows best practices (https://github.com/cake-build/website/issues/1280) as help for users to decide if they want to use an addin or not. Now imagine our official guideline is to not support all runners, then we would promote addins which might not run at all on a specific runner, over addins which would run perfectly fine. That doesn't make sense.

> The other way around is worse (IMO): Start with `netstandard2.0` and `net461` and later discover that moving to `net5.0` is harder and not do it and abandon the addin.

.NET 5 implements .NET Standard 2.0, so the API is there for 100%. Not targeting `net5.0` doesn't mean the addin won't run there. And it should be definitely best practice to test addins that they run on .NET 5 (which doesn't mean you should also mulit-target to `net5.0`). 

The only advantage mulit-targeting `net5.0` would bring is that it would allow to use newer features of the framework, but always come with the cost of conditional code, as long as all runners and platforms are supported (which should be best practices as explained above). If an author wants to go this route, that's perfectly fine, as it is perfectly fine to multi-target to other targets (and sometimes also is required from a technical point of view). But it should not be best practice to at least target `net5.0`.

> 1. These are just recommendations and as we know very few addin authors follow the recommendations... We have the data to prove it: As of December 20th 2020, only `25` out of `272` (10%) addins follow the recommendation of targeting both `netstandard2.0` and `net461` ([source](https://github.com/cake-build/website/pull/1400#issue-543169745)). So I expect the impact of recommending `net5.0` to be low.

The thing here is that recommendations changed a few times, also officially from Microsoft. In most cases targeting just `netstandard2.0` works also perfectly fine. But depending on what API you use there might be issues with .NET 4.6.1, therefore now the suggestion to multi-target to `netstandard2.0` and `net461` . But if an addin works fine with .NET 4.6.1 and `netstandard2.0` then there's not much reason to multi-target. 
Guidelines are never black and white. But they should provide sensible defaults which are best in most cases. There are always exceptions.

> 2. I do see value as I said in my previous comment, in keeping the projects active and healthy. Putting my build engineer hat on now 🧢 I'm more likely to take a dependency on an addin that had the last commit 3 months ago, than if the last commit was 3 years ago. If I see someone updating their addins to keep up with the latest versions of Cake and/or .NET, it sends me a signal that the addin is still active and the maintainer is engaged.

So, assuming an addin does what it should do, works on all runners and platforms, you would suggest to enforce change for the sake of change just to show that someone still maintains it? Doesn't make sense to me at all. 

> > Promoting multi-targeting to net5.0 [...] and adds the risk that they start to only target `net5.0`
> 
> You say that like it's a bad thing. IMHO we should be encouraging devs to move forward and leave the .NET Framework runner behind, leave the CoreCLR runner behind, and make the jump to the future: The dotnet tool running on .NET 5.

Yes, it is a **very** bad thing when we push addin authors to not support all runners and platforms which we decided to support with Cake itself. There are reason why we decided to support this platforms. This can change in the future, but it should come from the core as a decision for the whole ecosystem and not be driven by pushing addin authors in a specific direction.
Also you don't help them if for example Cake says .NET Framework is supported, but Cake says Addin authors should not support .NET Framework, then users will just complain and fill support issues for addins.

> At the end of the day, we're talking about build tools... Not applications, and as you know, one can easily have a build pipeline running on .NET 5, but building applications that still target .NET Framework 4.0 or even older... i.e. The build pipeline doesn't have to be as old as the application being built... It can evolve much faster than apps, which leaves little reason not to upgrade it.

We still have 3.5k downloads per day for .NET Framework runner on NuGet (The ones behind caching proxies not counted, which are very sensible to this). There are very valid reasons why existing build pipelines can't be easily moved to .NET Tool. It's absolutely OK to make a decision at some point that Cake no longer support .NET Framework, but to enforce this change through addins is the wrong way.

> Even Microsoft dev teams (with paid employees) dropped `netstandard2.0` moving forward (e.g. EF Core), so I don't think it's reasonable to expect unpaid volunteers (like addin authors) to keep supporting it indefinitely.

Again a big difference. See arguments above about addins being bound to Cake, `netstandard2.0` running perfectly fine on .NET 5.0 and user experience.

**Comment by augustoproiete** [comment]
@pascalberger and I decided we'll agree to disagree on this one, and revisit post Cake v1 release.
